sqlite_utils/gis.py

```diff
@@ -0,0 +1,144 @@
+import os
+from .db import Database, Table
+
+SPATIALITE_PATHS = (
+    "/usr/lib/x86_64-linux-gnu/mod_spatialite.so",
+    "/usr/local/lib/mod_spatialite.dylib",
+)
+
+
+def find_spatialite() -> str:
+    """
+    The ``find_spatialite()`` function searches for the `SpatiaLite <https://www.gaia-gis.it/fossil/libspatialite/index>`__ SQLite extension in some common places. It returns a string path to the location, or ``None`` if SpatiaLite was not found.
+
+    You can use it in code like this:
+
+    .. code-block:: python
+
+        from sqlite_utils import Database
+        from sqlite_utils.gis import find_spatialite
+
+        db = Database("mydb.db")
+        spatialite = find_spatialite()
+        if spatialite:
+            db.conn.enable_load_extension(True)
+            db.conn.load_extension(spatialite)
+    """
+    for path in SPATIALITE_PATHS:
+        if os.path.exists(path):
+            return path
+    return None
+
+
+def init_spatialite(db: Database, path: str) -> bool:
+    """
+    The ``init_spatialite`` function will load and initialize the Spatialite extension.
+    The ``path`` argument should be an absolute path to the compiled extension, which
+    can be found using ``find_spatialite``.
+
+    Returns true if Spatialite was successfully initialized.
+
+    .. code-block:: python
+
+        from sqlite_utils.gis import find_spatialite, init_spatialite
+
+        db = Database("mydb.db")
+        init_spatialite(db, find_spatialite())
+
+    If you've installed Spatialite somewhere unexpected (for testing an alternate version, for example)
+    you can pass in an absolute path:
+
+    .. code-block:: python
+
+        .. code-block:: python
+
+        from sqlite_utils.gis import init_spatialite
+
+        db = Database("mydb.db")
+        init_spatialite(db, "./local/mod_spatialite.dylib")
+
+    """
+    db.conn.enable_load_extension(True)
+    db.conn.load_extension(path)
+    # Initialize SpatiaLite if not yet initialized
+    if "spatial_ref_sys" in db.table_names():
+        return False
+    cursor = db.execute("select InitSpatialMetadata(1)")
+    result = cursor.fetchone()
+    return result and bool(result[0])
+
+
+def add_geometry_column(
+    table: Table,
+    geometry_type: str,
```

Comment:
I'm trying to figure out the list of valid geometry type values, which is surprisingly hard!

I eventually found a list on http://www.gaia-gis.it/gaia-sins/spatialite-sql-5.0.1.html by searching for `AddGeometryColumn` - that documentation says that valid values are:

- `POINT`, `POINTZ`, `POINTM`, `POINTZM`
- `LINESTRING`, `LINESTRINGZ`, `LINESTRINGM`, `LINESTRINGZM`
- `POLYGON`, `POLYGONZ`, `POLYGONM`, `POLYGONZM`
- `MULTIPOINT`, `MULTIPOINTZ`, `MULTIPOINTM`, `MULTIPOINTZM`
- `MULTILINESTRING`, `MULTILINESTRINGZ`, `MULTILINESTRINGM`, `MULTILINESTRINGZM`
- `MULTIPOLYGON`, `MULTIPOLYGONZ`, `MULTIPOLYGONM`, `MULTIPOLYGONZM`
- `GEOMETRYCOLLECTION`, `GEOMETRYCOLLECTIONZ`, `GEOMETRYCOLLECTIONZM`, `GEOMETRYCOLLECTIONZM`
- `GEOMETRY`, `GEOMETRYZ`, `GEOMETRYM`, `GEOMETRYZM`

I was going to suggest maybe providing constants for these and requiring that the argument was one of those constants... but with that many options I'm not sure that would be better than sticking with strings, and letting SpatiaLite itself raise an exception on an invalid value!

Reply:
I wonder if `GEOMETRY` would be a reasonable default value for `add_geometry_column(geometry_type=)`? I believe that's a column which can hold any type of geometry value - then we could allow users who want to be more restrictive to use `geometry_type="POINT"` or similar.

Reply:
Yeah, I had to dig for that, too. And those types are case-insensitive, as far as I can tell. I'm happy to provide constants, but if you pass in `"point"` as your geometry type, Spatialite won't complain.

Reply:
I'm fine to take out the default column name. Using `geometry` was my effort to standardize on GeoJSON, but I've seen plenty of examples that don't. Then you could have a similar signature, something like this:

```python
add_geometry_column(db["locations"], "geometry", "POINT")
create_spatial_index(db["locations"], "geometry")
```